container/container.go

```diff
@@ -138,52 +271,186 @@ func (c *DockerContainer) FindContainer(namespace string) (types.ContainerJSON,
-	return presenceHandling(err)
+	_, _, err := c.client.ServiceInspectWithRaw(context.Background(), c.namespace(namespace), types.ServiceInspectOptions{})
+	if err != nil && !client.IsErrNotFound(err) {
+		return false, err
+	}
+	return !client.IsErrNotFound(err), nil
 }
 
-// presenceHandling checks err to see if it's a Docker NotFound error and if not
-// it'll return the err back.
-func presenceHandling(err error) (bool, error) {
-	if err != nil && !docker.IsErrNotFound(err) {
-		return false, err
+// tasksError returns the error of matching tasks.
+func (c *DockerContainer) tasksError(namespace string) ([]string, error) {
+	tasks, err := c.client.TaskList(context.Background(), types.TaskListOptions{
+		Filters: filters.NewArgs(filters.KeyValuePair{
+			Key:   "label",
+			Value: "com.docker.stack.namespace=" + c.namespace(namespace),
+		}),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var errors []string
+	for _, task := range tasks {
+		if task.Status.Err != "" {
+			errors = append(errors, task.Status.Err)
+		}
+	}
+	return errors, nil
+}
+
+// waitForStatus waits for the container to have the provided status. Returns error as soon as possible.
+func (c *DockerContainer) waitForStatus(namespace string, status Status) error {
+	curstatus, err := c.serviceStatus(namespace)
+	if err != nil {
+		return err
+	}
+
+	for curstatus != status {
+		time.Sleep(100 * time.Millisecond)
+
+		tasksErrors, err := c.tasksError(namespace)
+		if err != nil {
+			return err
+		}
+		if len(tasksErrors) > 0 {
+			return errors.New(strings.Join(tasksErrors, ", "))
+		}
+
+		curstatus, err = c.serviceStatus(namespace)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *DockerContainer) createSharedNetwork() error {
+	// check if already exist
+	network, err := c.client.NetworkInspect(context.Background(), c.nsprefix, types.NetworkInspectOptions{})
+	if network.ID != "" {
+		c.sharedNetworkID = network.ID
+		return nil
+	}
+	if err != nil && !client.IsErrNotFound(err) {
+		return err
+	}
+
+	// Create the new network needed to run containers.
+	resp, err := c.client.NetworkCreate(context.Background(), c.nsprefix, types.NetworkCreate{
+		CheckDuplicate: true,
+		Driver:         "overlay",
+		Labels: map[string]string{
+			"com.docker.stack.namespace": c.nsprefix,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	c.sharedNetworkID = resp.ID
+	return nil
+}
+
+// readDockerIgnoreFile reads excluded file list from .dockerignore file.
+func readDockerIgnoreFile(path string) ([]string, error) {
+	f, err := os.Open(filepath.Join(path, ".dockerignore"))
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	return dockerignore.ReadAll(f)
+}
+
+func (c *DockerContainer) deletePendingContainer(namespace string, maxGraceTime time.Time) error {
+	var (
+		id  string
+		err error
+	)
+	for start := time.Now(); start.Before(maxGraceTime); time.Sleep(100 * time.Millisecond) {
```

Comment:
Does `start` values change over time? I don't see anywhere where `start` is re-assigned.